src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -7,13 +7,212 @@
+    cq = ibv_create_cq(id->verbs, 1, NULL, NULL, 0);
+    if (cq == NULL) {
+        ucs_error("ibv_create_cq() failed: %m");
+        status =  UCS_ERR_IO_ERROR;
+        goto out;
+    }
+
+    /* Create a dummy UD qp */
+    memset(&qp_init_attr, 0, sizeof(qp_init_attr));
+    qp_init_attr.send_cq = cq;
+    qp_init_attr.recv_cq = cq;
+    qp_init_attr.qp_type = IBV_QPT_UD;
+    qp_init_attr.cap.max_send_wr  = 2;
+    qp_init_attr.cap.max_recv_wr  = 2;
+    qp_init_attr.cap.max_send_sge = 1;
+    qp_init_attr.cap.max_recv_sge = 1;
+
+    qp = ibv_create_qp(id->pd, &qp_init_attr);
+    if (qp == NULL) {
+        ucs_error("failed to create a dummy ud qp. %m");
+        status = UCS_ERR_IO_ERROR;
+        goto out_destroy_cq;
+    }
+
+    ucs_debug("created ud QP %p with qp_num: 0x%x and cq %p on rdmacm_id %p",
+              qp, qp->qp_num, cq, id);
+
+    *cq_p = cq;
+    *qp_p = qp;
+
+    return UCS_OK;
+
+out_destroy_cq:
+    ibv_destroy_cq(cq);
+out:
+    return status;
+}
+
+ucs_status_t
+uct_rdamcm_cm_ep_set_remaining_conn_param(struct rdma_conn_param *conn_param,
+                                           const uct_rdmacm_priv_data_hdr_t *hdr,
+                                           uct_rdmacm_cm_ep_t *cep)
+{
+    ucs_status_t status;
+    struct ibv_qp *qp;
+    struct ibv_cq *cq;
+
+    /* create a dummy qp in order to get a unique qp_num to provide to librdmacm */
+    status = uct_rdmacm_create_dummy_cq_qp(cep->id, &cq, &qp);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    cep->cq                      = cq;
+    cep->qp                      = qp;
+
+    conn_param->qp_num           = qp->qp_num;
+    conn_param->private_data_len = sizeof(*hdr) + hdr->length;
+    return UCS_OK;
+}
+
+static ucs_status_t uct_rdamcm_cm_ep_init_client_ep(uct_rdmacm_cm_ep_t *cep,
+                                                    const uct_ep_params_t *params)
+{
+    ucs_status_t status;
+
+    cep->wireup.client.connect_cb = params->sockaddr_connect_cb.client;
+
+    if (rdma_create_id(cep->cm->ev_ch, &cep->id, cep, RDMA_PS_TCP)) {
+        ucs_error("rdma_create_id() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err;
+    }
+
+    if (rdma_resolve_addr(cep->id, NULL, (struct sockaddr *)params->sockaddr->addr,
+                          1000/* TODO */)) {
+        ucs_error("rdma_resolve_addr() failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto err_destroy_id;
+    }
+
+    return UCS_OK;
+
+err_destroy_id:
+    rdma_destroy_id(cep->id);
+err:
+    return status;
+}
+
+static ucs_status_t uct_rdamcm_cm_ep_init_server_ep(uct_rdmacm_cm_ep_t *cep,
+                                                    const uct_ep_params_t *params)
 {
     return UCS_ERR_NOT_IMPLEMENTED;
 }
 
+ucs_status_t uct_rdmacm_cm_ep_disconnect(uct_ep_h ep, unsigned flags)
+{
+    uct_rdmacm_cm_ep_t *cep = ucs_derived_of(ep, uct_rdmacm_cm_ep_t);
+    struct sockaddr *remote_addr = rdma_get_peer_addr(cep->id);
```

Comment:
can put `rdma_get_peer_addr` call directly into `ucs_sockaddr_str` to avoid overhead for successful flow in release build